x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -57,12 +59,45 @@
 import java.util.stream.Collectors;
 
 import static org.elasticsearch.index.query.QueryBuilders.boolQuery;
+import static org.elasticsearch.xpack.ql.index.IndexResolver.UNMAPPED;
 import static org.elasticsearch.xpack.ql.util.ActionListeners.map;
 import static org.elasticsearch.xpack.ql.util.StringUtils.WILDCARD;
 
 public class EsqlSession {
 
     private static final Logger LOGGER = LogManager.getLogger(EsqlSession.class);
+    public static final BiFunction<String, Map<String, FieldCapabilities>, InvalidMappedField> SPECIFIC_VALIDITY_VERIFIER = (
+        name,
+        types) -> {
+        boolean hasUnmapped = types.containsKey(UNMAPPED);
+        boolean hasTypeConflicts = types.size() > (hasUnmapped ? 2 : 1);
+        String metricConflictsTypeName = null;
+        boolean hasMetricConflicts = false;
+
+        if (hasTypeConflicts == false) {
+            for (Map.Entry<String, FieldCapabilities> type : types.entrySet()) {
+                if (UNMAPPED.equals(type.getKey())) {
+                    continue;
+                }
+                if (type.getValue().metricConflictsIndices() != null && type.getValue().metricConflictsIndices().length > 0) {
+                    hasMetricConflicts = true;
+                    metricConflictsTypeName = type.getKey();
+                    break;
+                }
+            }
+        }
+
+        if (hasMetricConflicts) {
+            StringBuilder errorMessage = new StringBuilder();
+            errorMessage.append(
+                "mapped as different metric types in indices: ["
+                    + String.join(", ", types.get(metricConflictsTypeName).metricConflictsIndices())
+                    + "]"
+            );
+            return new InvalidMappedField(name, errorMessage.toString());
+        }
+        return null;
```

Comment:
Suggest a small change in readability around the code flow to avoid having two return statements:
```
InvalidMappedField result = null;

if (hasMetricConflicts) {
 ....
result = new InvalidMappedField()
}
return results;
```
or promote the errorMessage above the if then do:
`return errorMessage != null  ? new InvalidMappedField(name, errorMessage.toString()) : null);`